src/gap-skipper.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * @file gap-skipper.js
+ */
+import Ranges from './ranges';
+import videojs from 'video.js';
+
+/**
+ * the gap skipper object handles all scenarios
+ * where the player runs into the end of a buffered
+ * region and there is a buffered region ahead.
+ * It then handles the skipping behavior.
+ *
+ * @class GapSkipper
+ */
+export default class GapSkipper {
+  /**
+   * Represents a GapSKipper object.
+   * @constructor
+   * @param {object} options an object that includes the tech and settings
+   */
+  constructor(options) {
+    if (!options.tech.options_.playerId) {
+      return;
+    }
+
+    this.player_ = videojs(options.tech.options_.playerId);
+    this.tech_ = options.tech;
+    this.consecutiveUpdates = 0;
+    this.lastRecordedTime = null;
+    this.timer_ = null;
+
+    if (options.debug) {
+      this.logger_ = videojs.log.bind(videojs, '<gap-skipper>');
+    }
+    this.logger_('<initialize>');
+
+    // The purpose of this function is to emulate the "waiting" event on
+    // browsers that does not emit it when they are stalled waiting for
+    // more data
```

Comment:
I don't think `waiting` is actually the event we're looking for here. My reading of the spec says [waiting](https://html.spec.whatwg.org/multipage/embedded-content.html#event-media-waiting) is reserved for scenarios where the user or user agent has triggered a condition that requires buffering before playback but there's no evidence that buffering won't be 100% successful. For instance, seeking out of buffer would trigger waiting because it's an intentional decision to move into a state that requires buffering. 

[stalled](https://html.spec.whatwg.org/multipage/embedded-content.html#event-media-stalled) indicates that buffering is being attempted but no progress is being made. This sounds closer to the scenario we're dealing with ("I'm trying to buffer but failing!") but still not right.

[A related issue](https://github.com/w3c/media-source/issues/88) was recently filed against the media sources spec. If the application developer (us) is doing the content fetching, are `progress`, `waiting`, and `stalled` events well-defined? I lean towards "no" which makes me think we shouldn't overload an existing event for this functionality. All this is an extraordinarily long-winded way of saying, I think we should make our own event, possibly trigger it in response to `waiting` and `stalled`, and use that new event for gap skips.